### ISSUE TITLE
Unsigned auxiliar scripts of Wazuh Windows Agent WPK could lead to AV/EDR alarms

### ISSUE THREAD

**Issue by jnasselle** [wazuh/wazuh-packages#2146]
|Wazuh version|Install type|Action performed|Platform|
|---|---|---|---|
| up to master | Windows Agent WPK | Upgrade | Windows |

# Description

A community from our community reported that his EDR/AV show alerts on the next Wazuh Wazuh Agent WPK aux files:

- https://github.com/wazuh/wazuh/blob/4.4/src/win32/upgrade.bat
- https://github.com/wazuh/wazuh/blob/4.4/src/win32/do_upgrade.ps1

Despite both WPK bundle and Wazuh MSI inside being signed/verified, those two files could cause errors during upgrades

**Comment by lchico** [comment]
# Update
I've started working on this and am looking to replicate the issue

**Comment by lchico** [comment]
# Update
2024/07/11 -  I tried to sign both scripts, but the update.bat could not be signed. Now, I am attempting to convert the batch script to a PowerShell script.
2024/07/12 - I successfully generated the signed MSI and WPK files locally, but encountered issues during the update process. I'm investigating the cause of the "Send upgrade command error."
The command used for testing:
```
/var/ossec/bin/agent_upgrade -a 006 -f $(pwd)/test.wpk -x upgrade.ps1

Upgrading...

Failed upgrades:
	Agent 006 status: Send upgrade command error

```
Additionally, I worked on addressing the issue https://github.com/wazuh/wazuh-agent-packages/issues/50 by globally setting the error action to "Stop":
```
$ErrorActionPreference = "Stop"
```

I have run [this workflow](https://github.com/wazuh/wazuh-agent-packages/actions/runs/9930666238/job/27429658065) to test the fix.

**Comment by lchico** [comment]
# Update

Initially, I believed the issue was that we needed to sign the update.bat and do_update.ps1 files when generating the MSI file. However, I now realize that this should be done during the wpk generation process or the signing process, and then these files should be copied when generating the wpk file. Additionally, I encountered an issue where I couldn't execute an upgrade.ps1 script using the upgrade command. This script merely creates a simple file during the upgrade process, this never happens.

During local testing, one of the problems I encountered was that the custom wpk file I tested did not have the correct certificate. I resolved this by updating the active response on the agent to **disable CA verification**. As a result, I confirmed that the _upgrade_, _do_upgrade_, and _binary_ files are uncompressed properly.

However, the output on the manager still indicates:
```Agent 006 status: Send upgrade command error.```

Regarding [wazuh/wazuh-agent-packages#50](https://github.com/wazuh/wazuh-agent-packages/issues/50), globally setting the error action to "Stop" did not work. Additionally, I reviewed the workflow and attempted to run it, but encountered issues when trying to execute the allocator. I faced similar problems with the workflow stopping, as observed in [this execution](https://github.com/wazuh/wazuh-agent-packages/actions/runs/9947500393/job/27480505482).

To investigate further, I performed tests locally and identified two potential issues:
####  step: Create MSI package
```bash
       ${{ env.SSH_COMMAND }} "cd C:/win-agent-base/wazuh-local-src/src/win32/; ./generate_wazuh_msi.ps1 -MSI_NAME ${{ env.PKG_NAME }}.msi -SIGN yes 2>&1 > C:/generation.log"
          ${{ env.SCP_COMAND }} ${{ env.ansible_user }}@${{ env.ansible_host }}:"C:/generation.log" /tmp/
          cat /tmp/generation.log
          if [ $(grep 'SignTool Error' /tmp/generation.log) ] ; then
            echo "The generation and sign could not be completed."
            exit 1
          else
```
1. It's possible that this command does not send errors to standard output. To address this, we should modify it as follows:
```
       ${{ env.SSH_COMMAND }} "cd C:/win-agent-base/wazuh-local-src/src/win32/; ./generate_wazuh_msi.ps1 -MSI_NAME ${{ env.PKG_NAME }}.msi -SIGN yes 2>&1 > C:/generation.log"
```
**Note:** Update the end of the command to include **2>&1 > C:/generation.log"** or is just the next option.

2. During local testing, I noticed that grep did not match any words, possibly due to the file having Windows encoding. We may need to consider converting the file's encoding or finding an alternative solution.

**Comment by lchico** [comment]
# Update

While debugging the upgrade process, I found that the issues arose because Windows does not recognize `.ps1` files as executable.
```log
2024/07/16 12:10:09 wazuh-agent: WARNING: No root CA defined to verify file 'incoming\wazuh-agent-test-v3.wpk'.
2024/07/16 12:10:09 wazuh-agent: ERROR: at wm_exec(): CreateProcess(193): %1 no es una aplicacion Win32 valida.
2024/07/16 12:10:09 wazuh-modulesd:agent-upgrade: ERROR: (8135): At upgrade: Error executing command [upgrade\upgrade.ps1]
```

Therefore, I would like to confirm this. I found [this link](https://stackoverflow.com/questions/10137146/is-there-a-way-to-make-a-powershell-script-work-by-double-clicking-a-ps1-file#:~:text=Or%20if%20you%20want%20all%20PS1%20files%20to%20work%20the%20way%20VBS%20files%20do%2C%20you%20can%20edit%20the%20registry%20key%20at%3A) which suggests that changing some registry settings may allow the execution of PS1 scripts.

**Comment by lchico** [comment]
# Update

| Possible Solution                                                                 | Description                                                                                                                   | Notes                                                   |
|--------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------|
| 1. Update C code to call PowerShell                                      | Modify the existing C code to invoke the PowerShell script directly.                                                           | I invested some time trying to make it work but without success.                   |
| 2. Change Windows registry to recognize PowerShell script                | Modify the Windows registry to allow the execution of PowerShell scripts.                                                      | Not recommended due to potential security risks.       |
| 3. Create a C++ script to call the PowerShell script                     | Develop a new C++ script specifically to call and execute the PowerShell script.                                               | I tried to do a simple proof of concept but couldn't make it work.           |

**Comment by vikman90** [comment]
@lchico Thanks!

Regarding option 1, this is the reason why the WPK upgrade requires an executable file:
- [wm_agent_upgrade_com.c](https://github.com/wazuh/wazuh/blob/3b44b3b6d608bb4a90353a74a1b32684277cebd9/src/wazuh_modules/agent_upgrade/agent/wm_agent_upgrade_com.c#L364-L368https://github.com/wazuh/wazuh/blob/3b44b3b6d608bb4a90353a74a1b32684277cebd9/src/wazuh_modules/agent_upgrade/agent/wm_agent_upgrade_com.c#L364-L368)

It expects an executable file (parameter `installer` or option `-x`).

I would discard option 2. Therefore, we only have option 3 left.

## Issue blocked

We must decide:

- If we cannot sign BAT files and this signature is critical to getting approval from antivirus and security tools, we will have to translate `upgrade.bat` to C code.
- If we are going to continue making changes to 5.x in WPK upgrades, it may be worth waiting for this and taking it into account in the redesign.

**Comment by vikman90** [comment]
@lchico @MarcelKemp we choose the second option: do not sign `upgrade.bat`.

We're going to redesign the agent upgrade in upcoming versions, so let's sign `do_upgrade.ps1` only, since it is a simple task, as I understand it.

Regards!

**Comment by lchico** [comment]
# Update
Continue working on this task. Update the test/2146-windows_agent_package_signing_error_false_positive branch with a possible solution. I'll need to test it.

**Comment by lchico** [comment]
# Update
After several tries, I couldn't sign the `do_upgrade.ps1` under Linux to make it compatible with Windows. (I mention this because we generate the WPK under Linux, downloading the signed MSI file and adding the `upgrade.bat` and `do_upgrade.ps1` from the selected branch)

![signed](https://github.com/user-attachments/assets/0385d72c-a432-42df-ab8e-2d370411edbe)

| Option | Pros | Cons |
| ------ | ---- | ---- |
| **1. Signing `do_upgrade.ps1` on Linux as part of the WPK process** | - I thought this would be the best option. | - I spent several hours without success.  Given my expertise, I'm not sure if this is possible.|
| **2. Push the signed `do_upgrade.ps1` to S3 when generating the MSI** | - Simplifies the signing process. | - Requires additional steps to upload and manage files on S3. <br> - Slightly more complex workflow. |
| **3. Push the signed `do_upgrade.ps1` to the source** | - Guarantees the `do_upgrade.ps1` is always signed. <br> - Simplifies the WPK generation process on Linux. | - Involves maintaining and managing the signed script in the source whenever we make any changes to `do_upgrade.ps1`. |

**Note:** The simplest option would be the third one.

**Comment by lchico** [comment]
# Update 
I successfully added the signed do_upgrade.ps1 script to the MSI file and successfully extracted it on a Linux environment. Local testing was successful. The next step is to implement this change in the workflows.

**Comment by lchico** [comment]
# Update

TL;DR: Solutions were pushed to the do_upgrade.ps1 script, which has been signed. ([Proposal number 3](https://github.com/wazuh/wazuh-packages/issues/2146#issuecomment-2253686612))

After some back and forth, I was able to generate the MSI with the do_upgrade.ps1 script signed. The idea was to extract the script from the MSI and add it to the WPK. That's when the problem started.

1. When I extracted it on Linux using the **msiextract** tool, the file had the signed block at the end, but it wasn't recognized by Windows.
![image](https://github.com/user-attachments/assets/953a6531-7e1f-409f-ad87-23467e9ba30a)


2. Then I tried to compress the **do_upgrade.ps1** and **embed** it into the **MSI**. However, when I tried to decompress the file, I got the following error:
![Screenshot from 2024-07-30 16-41-14](https://github.com/user-attachments/assets/7590bd76-c279-41c8-93aa-e9ae41047491)
It seems this method is somewhat unstable.

As mentioned, I ended up pushing the signed do_upgrade script, and it is working as expected. To get the do_upgrade.ps1 signed, I used the following commit: [commit link](https://github.com/wazuh/wazuh/commit/b4ef4c31dd93c8bfcdc97289780477309dc45863), which creates the MSI. The [workflow run](https://github.com/wazuh/wazuh-agent-packages/actions/runs/10167469402/job/28119944153) embeds the script inside the MSI package, and I used 7z on Windows to extract the file.

```git
diff --git a/src/win32/wazuh-installer.wxs b/src/win32/wazuh-installer.wxs
--- a/src/win32/wazuh-installer.wxs
+++ b/src/win32/wazuh-installer.wxs
@@ -468,7 +462,11 @@
                     <Directory Id="RIDS" Name="rids" />
                     <Directory Id="SYSCHECK" Name="syscheck" />
                     <Directory Id="INCOMING" Name="incoming" />
-                    <Directory Id="UPGRADE" Name="upgrade" />
+                    <Directory Id="UPGRADE" Name="upgrade">^M
+                        <Component Id="DO_UPGRADE.PS1" DiskId="1" Guid="37D0E724-F330-4897-8B2E-F751D84AD61C" Permanent="yes">^M
+                            <File Id="DO_UPGRADE.PS1" Name="do_upgrade.ps1" Source="do_upgrade.ps1" KeyPath="yes" />^M
+                        </Component>^M
+                    </Directory>^M
                 </Directory>
             </Directory>
             <Directory Id="ProgramMenuFolder">
@@ -617,6 +615,7 @@
         <Feature Id="MainFeature" Title="Wazuh Agent" Description="Install the Wazuh Agent program files" ConfigurableDirectory="APPLICATIONFOLDER" InstallDefault="local" TypicalDefault="install" AllowAdvertise="no" Absent="disallow">
             <ComponentRef Id="REGISTRY_INSTALL_DIR" />
             <ComponentRef Id="AGENT_AUTH.EXE" />
+            <ComponentRef Id="DO_UPGRADE.PS1" />^M
             <ComponentRef Id="SYSCOLLECTOR_NORM_CONFIG" />
             <ComponentRef Id="SYSCOLLECTOR_DLL" />
             <ComponentRef Id="FIMDB_DLL" />
```

**Note:** If we implement this solution, we should consider the signature's expiration date.


![image](https://github.com/user-attachments/assets/dd9e8c76-ff36-45c0-975f-5a0fe98dfd84)


It's possible that by dedicating more effort, I could find a way to compress the PS1 file on Windows so that the signature remains valid when uncompressed on Linux.

**Comment by vikman90** [comment]
We have two options to proceed with the signing:
1. Build the WPK package along with the MSI, on the machine that handles the signing.
2. Have the DigiCert allocator machine generate the MSI and PS1, upload both artifacts to S3, and then generate the WPK as usual.

**Comment by vikman90** [comment]
The solution is to remove the PS1/BAT scripts in a next version.
